google/cloud/storage/blob.py

```diff
@@ -468,9 +474,21 @@ def generate_signed_url(
             raise ValueError("'version' must be either 'v2' or 'v4'")
 
         quoted_name = _quote(self.name, safe=b"/~")
-        resource = "/{bucket_name}/{quoted_name}".format(
-            bucket_name=self.bucket.name, quoted_name=quoted_name
-        )
+
+        if virtual_hosted_style_endpoint and version == "v4":
+            storage_uri = api_access_endpoint.split("://")
+            api_access_endpoint = (
+                storage_uri[0] + "://" + self.bucket.name + "." + storage_uri[1]
+            )
```

Comment:
I'd rather have
`api_access_endpoint = "{}://{}.{}".format(storage_uri[0], self.bucket.name, storage_uri[1])`
although it's a matter of personal preference.